README.md
# grunt-phantomcss

> Automate CSS regression testing with PhantomCSS

## Notice

**This is a fork of the original (presumably discontinued) repository of [grunt-phantomcss](https://github.com/chrisgladd/grunt-phantomcss). Currently this version here is untagged and unreleased on npm. However, you can install and use this version:**

Add this to your `package.json`:

    "grunt-phantomcss": "git://github.com/anselmh/grunt-phantomcss.git",

or, alternatively, type this into your command line interface:

    npm i --save-dev git://github.com/anselmh/grunt-phantomcss.git

## CHANGELOG

* Added `mismatchTolerance` tolerance (`mismatchTolerance: 0.0.5` is default) (thx [fidgety](https://github.com/chrisgladd/grunt-phantomcss/pull/17))
* Use PhantomCSS v0.7.1
* Update dependencies
* Remove grunt as peerDependency causing multiple issues with peerPackages
* Fix post-script `bower install`

## TO DO:

* Add timeout as an optional argument (currently can fail on large testing areas with a timeout)

----

## Getting Started
This plugin requires Grunt `~0.4.1`

If you haven't used [Grunt](http://gruntjs.com/) before, be sure to check out the [Getting Started](http://gruntjs.com/getting-started) guide, as it explains how to create a [Gruntfile](http://gruntjs.com/sample-gruntfile) as well as install and use Grunt plugins. Once you're familiar with that process, you may install this plugin with this command:

```shell
npm install grunt-phantomcss --save-dev
```

Once the plugin has been installed, it may be enabled inside your Gruntfile with this line of JavaScript:

```js
grunt.loadNpmTasks('grunt-phantomcss');
```

## The "phantomcss" task

### Overview
In your project's Gruntfile, add a section named `phantomcss` to the data object passed into `grunt.initConfig()`.

```js
grunt.initConfig({
  phantomcss: {
    options: {},
    your_target: {
      options: {
        screenshots: 'test/visual/screenshots/',
        results: 'results/visual/',
        viewportSize: [1280, 800],
        mismatchTolerance: 0.05
      },
      src: [
        'test/visual/**/*.js'
      ]
    }
  }
});
```

### Options

#### src
Type: `String|Array`

The test files to run.

#### options.screenshots
Type: `String`
Default: `'./screenshots'`

The screenshots directory where test fixtures (comparison screenshots) are stored. Baseline screenshots will be stored here on the first run if they're not present.

#### options.results
Type: `String`
Default: `'./results'`

The directory to store source, diff, and failure screenshots after tests.

#### options.viewportSize
Type: `Array`
Default: `[1280, 800]`

The viewport size to test the site in `[width, height]` format. Useful when testing responsive layouts.

#### options.logLevel
Type: `String`
Default: `error`

The CasperJS log level. See [CasperJS: Logging](http://casperjs.readthedocs.org/en/latest/logging.html) for details.


### Usage Examples

#### Basic visual tests
Run tests in `test/visual/` against comparison screenshots stored in `test/visual/screenshots/`, and put the resulting screenshots in `results/visual/`

```js
grunt.initConfig({
  phantomcss: {
    options: {
      screenshots: 'test/visual/screenshots/',
      results: 'results/visual/'
    },
    src: [
      'test/visual/**/*.js'
    ]
  }
});
```

#### Responsive layout testing
Run tests in `test/visual/` against comparison screenshots for destop and mobile.

```js
grunt.initConfig({
  phantomcss: {
    desktop: {
      options: {
        screenshots: 'test/visual/desktop/',
        results: 'results/visual/desktop',
        viewportSize: [1024, 768]
      },
      src: [
        'test/visual/**.js'
      ]
    },
    mobile: {
      options: {
        screenshots: 'test/visual/mobile/',
        results: 'results/visual/mobile',
        viewportSize: [320, 480]
      },
      src: [
        'test/visual/**.js'
      ]
    }
  },
});
```

#### Sample test file

Test files should do the following:
* Start CasperJS with the URL you want to test
* Manipulate the page in some way
* Take screenshots

```javascript
casper.start('http://localhost:3000/')
.then(function() {
  phantomcss.screenshot('#todo-app', 'Main app');
})
.then(function() {
  casper.fill('form.todo-form', {
    todo: 'Item1'
  }, true);

  phantomcss.screenshot('#todo-app', 'Item added');
})
.then(function() {
  casper.click('.todo-done');

  phantomcss.screenshot('#todo-app', 'Item checked off');
});
```

You can also load a local file by specifying a path (relative to the Gruntfile):

```javascript
casper.start('build/client/index.html')
.then(function() {
  // ...
});
```

### Multiple Test Files
Your first test file should use ```casper.start```

```javascript
casper.start('http://localhost:3000/')
.then(function() {
  phantomcss.screenshot('#todo-app', 'Main app');
})
.then(function() {
  casper.fill('form.todo-form', {
    todo: 'Item1'
  }, true);

  phantomcss.screenshot('#todo-app', 'Item added');
});

```
Subsequent files should call ```casper.then``` to continue the previous test.

```javascript
casper.then(function() {
  casper.click('.todo-done');

  phantomcss.screenshot('#todo-app', 'Item checked off');
});
```
You can also use ```casper.thenOpen``` to load a new url and continue testing in subsequent files instead of ```casper.start```.


See the [CasperJS documentation](http://casperjs.readthedocs.org/en/latest/modules/casper.html) and the [PhantomCSS documentation](https://github.com/Huddle/PhantomCSS) for more information on using CasperJS and PhantomCSS.


## Contributing
In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality. Lint and test your code using [Grunt](http://gruntjs.com/).

## Release History
* 2014-02-23   v0.2.2   Added multiple file example to README.md
* 2014-02-07   v0.2.1   Fixed ResembleJS path issue
* 2014-01-07   v0.2.0   Merged updates from Larry Davis
* 2013-10-24   v0.1.1   Added the ability to use an external server
* 2013-10-24   v0.1.0   Initial Release


CHANGELOG.md
# CHANGELOG

### HEAD
* Update dependencies
* Add `mismatchTolerance`as option
* Add `waitTimeout` as option

### v0.3.1
* Keep log level at "notice"
* Add time-grunt to measure build times
* Add jit-grunt to load tasks only when needed (+ improve performance)
* Change log level to notice in Gruntfile
* Document correct CasperJS log levels and pass them as arguments / option
* Added .editorconfig

### v0.3.0
* Added `mismatchTolerance` tolerance (`mismatchTolerance: 0.0.5` is default) (thx [fidgety](https://github.com/chrisgladd/grunt-phantomcss/pull/17))
* Use PhantomCSS v0.7.1
* Update dependencies
* Remove grunt as peerDependency causing multiple issues with peerPackages
* Fix post-script `bower install`


tasks/phantomcss.js
/*
 * grunt-phantomcss
 * https://github.com/anselmh/grunt-phantomcss
 *
 * Copyright (c) 2013 Chris Gladd
 * Copyright (c) since 2014 Anselm Hannemann
 *
 * Licensed under the MIT license.
 */

'use strict';

var path = require('path');
var tmp = require('temporary');
var phantomBinaryPath = require('phantomjs').path;
var runnerPath = path.join(__dirname, '..', 'phantomjs', 'runner.js');
var phantomCSSPath = path.join(__dirname, '..', 'bower_components', 'phantomcss');

module.exports = function(grunt) {
    grunt.registerMultiTask('phantomcss', 'CSS Regression Testing', function() {
        var done = this.async();

        // Variable object to set default values for options
        var options = this.options({
            screenshots: 'screenshots',
            results: 'results',
            viewportSize: [1280, 800],
            mismatchTolerance: 0.05,
            logLevel: 'warning' // debug | info | warning | error
        });

        // Timeout ID for message checking loop
        var messageCheckTimeout;

        // The number of tempfile lines already read
        var lastLine = 0;

        // The number of failed tests
        var failureCount = 0;

        // This is effectively the project root (location of Gruntfile)
        // This allows relative paths in tests, i.e. casper.start('someLocalFile.html')
        var cwd = process.cwd();

        // Create a temporary file for message passing between the task and PhantomJS
        var tempFile = new tmp.File();

        var deleteDiffScreenshots = function() {
            // Find diff/fail files
            var diffScreenshots = grunt.file.expand([
                path.join(options.screenshots, '*diff.png'),
                path.join(options.screenshots, '*fail.png')
            ]);

            // Delete all of 'em
            diffScreenshots.forEach(function(filepath) {
                grunt.file.delete(filepath);
            });
        };

        var cleanup = function(error) {
            // Remove temporary file
            tempFile.unlink();

            // Create the output directory
            grunt.file.mkdir(options.results);

            // Copy fixtures, diffs, and failure images to the results directory
            var allScreenshots = grunt.file.expand(path.join(options.screenshots, '**.png'));
            allScreenshots.forEach(function(filepath) {
                grunt.file.copy(filepath, path.join(options.results, path.basename(filepath)));
            });

            deleteDiffScreenshots();

            done(error || failureCount === 0);
        };

        var checkForMessages = function checkForMessages(stopChecking) {
            // Disable logging temporarily
            grunt.log.muted = true;

            // Read the file, splitting lines on \n, and removing a trailing line
            var lines = grunt.file.read(tempFile.path).split('\n').slice(0, -1);

            // Re-enable logging
            grunt.log.muted = false;

            // Iterate over all lines that haven't already been processed
            lines.slice(lastLine).some(function(line) {
                // Get args and method
                var args = JSON.parse(line);
                var eventName = args[0];

                // Debugging messages
                grunt.log.debug(JSON.stringify(['phantomjs'].concat(args)).magenta);

                // Call handler
                if (messageHandlers[eventName]) {
                    messageHandlers[eventName].apply(null, args.slice(1));
                }
            });

            // Update lastLine so previously processed lines are ignored
            lastLine = lines.length;

            if (stopChecking) {
                clearTimeout(messageCheckTimeout);
            }
            else {
                // Check back in a little bit
                messageCheckTimeout = setTimeout(checkForMessages, 100);
            }
        };

        var messageHandlers = {
            onFail: function(test) {
                grunt.log.writeln('Visual change found for ' + path.basename(test.filename) + ' (' + test.mismatch + '% mismatch)');
            },
            onPass: function(test) {
                grunt.log.writeln('No changes found for ' + path.basename(test.filename));
            },
            onTimeout: function(test) {
                grunt.log.writeln('Timeout while processing ' + path.basename(test.filename));
            },
            onComplete: function(allTests, noOfFails, noOfErrors) {
                if (allTests.length) {
                    var noOfPasses = allTests.length - failureCount;
                    failureCount = noOfFails + noOfErrors;

                    if (failureCount === 0) {
                        grunt.log.ok('All ' + noOfPasses + ' tests passed!');
                    }
                    else {
                        if (noOfErrors === 0) {
                            grunt.log.error(noOfFails + ' tests failed.');
                        }
                        else {
                            grunt.log.error(noOfFails + ' tests failed, ' + noOfErrors + ' had errors.');
                        }
                    }
                }
                else {
                    grunt.log.ok('Baseline screenshots generated in '+options.screenshots);
                    grunt.log.warn('Check that the generated screenshots are visually correct and delete them if they aren\'t.');
                }
            }
        };

        // Resolve paths for tests
        options.test = [];
        this.filesSrc.forEach(function(filepath) {
            options.test.push(path.resolve(filepath));
        });

        options.screenshots = path.resolve(options.screenshots);

        // Put failure screenshots in the same place as source screenshots, we'll move/delete them after the test run
        // Note: This duplicate assignment is provided for clarity; PhantomCSS will put failures in the screenshots folder by default
        options.failures = options.screenshots;

        // Pass necessary paths
        options.tempFile = tempFile.path;
        options.phantomCSSPath = phantomCSSPath;

        // Remove old diff screenshots
        deleteDiffScreenshots();

        // Start watching for messages
        checkForMessages();

        grunt.util.spawn({
            cmd: phantomBinaryPath,
            args: [
                runnerPath,
                JSON.stringify(options)
            ],
            opts: {
                cwd: cwd,
                stdio: 'inherit'
            }
        }, function(error, result, code) {
            // When Phantom exits check for remaining messages one last time
            checkForMessages(true);

            cleanup(error);
        });
    });
};


Gruntfile.js
/*
 * grunt-phantomcss
 * https://github.com/anselmh/grunt-phantomcss
 *
 * Copyright (c) 2013 Chris Gladd
 * Copyright (c) since 2014 Anselm Hannemann
 *
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function (grunt) {
    // Project configuration.
    grunt.initConfig({
        jshint: {
            all: [
                'Gruntfile.js',
                'tasks/phantomcss.js',
                'phantomjs/runner.js',
                '<%= nodeunit.tests %>',
            ],
            options: {
                jshintrc: '.jshintrc',
            },
        },

        // Before generating any new files, remove any
        // previously-created files.
        clean: {
            tests: ['tmp'],
            results: ['results'],
        },

        // Configuration to be run (and then tested).
        phantomcss: {
            visualTest: {
                options: {
                    screenshots: 'fixtures/screenshots/',
                    results: 'results/',
                    viewportSize: [1280, 800],
                    mismatchTolerance: 0.05,
                    logLevel: 'warning' // debug | info | warning | error
                },
                src: [
                    'fixtures/coffeemachine_test.js'
                ]
            }
        },

        // Unit tests.
        nodeunit: {
            tests: ['test/*_test.js'],
        },
    });

    // Timings for Grunt
    require('time-grunt')(grunt);

    // Load all tasks JIT
    require('jit-grunt')(grunt, {
        phantomcss: 'tasks/phantomcss.js'
    });

    // Whenever the "test" task is run, first clean the "tmp" dir,
    // then run this plugin's task(s), then test the result.
    grunt.registerTask('test', ['clean', 'phantomcss', 'nodeunit']);

    // By default, lint and run all tests.
    grunt.registerTask('default', ['jshint', 'test']);
};
